pimd/pim_autorp.c

```diff
@@ -232,79 +401,321 @@ static void autorp_rp_holdtime(struct event *evt)
+		if (buf_sz < *sz + rp_sz) {
+			if (PIM_DEBUG_AUTORP)
+				zlog_warn("%s: Failed to pack AutoRP discovery packet, buffer overrun, (%u < %u)",
+					  __func__, (uint32_t)buf_sz, (uint32_t)(*sz + rp_sz));
+			break;
+		}
+
+		if (PIM_DEBUG_AUTORP)
+			zlog_debug("%s: Add RP %pI4 (grpcnt=%u) to discovery message", __func__,
+				   &rp->addr, (uint32_t)grpcnt);
+
+		rpcnt++;
+
+		brp = (struct autorp_pkt_rp *)(buf + bsz);
+		bsz += sizeof(struct autorp_pkt_rp);
+
+		/* Since this is an in_addr, assume it's already the right byte order */
+		brp->addr = rp->addr.s_addr;
+		brp->pimver = AUTORP_PIM_V2;
+		brp->reserved = 0;
+		brp->grpcnt = grpcnt;
+
+		frr_each (pim_autorp_grppfix, &rp->grp_pfix_list, grp) {
+			bgrp = (struct autorp_pkt_grp *)(buf + bsz);
+			bsz += sizeof(struct autorp_pkt_grp);
+
+			bgrp->addr = grp->grp.u.prefix4.s_addr;
+			bgrp->masklen = grp->grp.prefixlen;
+			bgrp->negprefix = grp->negative;
+
+			if (PIM_DEBUG_AUTORP)
+				zlog_debug("%s: Add group %s%pFX for RP %pI4 to discovery message",
+					   __func__, (grp->negative ? "!" : ""), &grp->grp,
+					   &rp->addr);
+		}
+
+		/* Update the size with this RP now that it is packed */
+		*sz += bsz;
+	}
+
+	return rpcnt;
+}
+
+static size_t autorp_build_disc_packet(struct pim_autorp *autorp, uint8_t *buf, size_t buf_sz)
+{
+	size_t sz = 0;
+	struct autorp_pkt_hdr *hdr;
+
+	if (buf_sz >= AUTORP_HDRLEN) {
+		hdr = (struct autorp_pkt_hdr *)buf;
+		hdr->version = AUTORP_VERSION;
+		hdr->type = AUTORP_DISCOVERY_TYPE;
+		hdr->holdtime = htons(autorp->discovery_holdtime);
+		hdr->reserved = 0;
+		sz += AUTORP_HDRLEN;
+		hdr->rpcnt = autorp_build_disc_rps(autorp, buf + sizeof(struct autorp_pkt_hdr),
+						   (buf_sz - AUTORP_HDRLEN), &sz);
+		if (hdr->rpcnt == 0)
+			sz = 0;
+	}
+	return sz;
+}
+
+static void autorp_send_discovery(struct event *evt)
+{
+	struct pim_autorp *autorp = EVENT_ARG(evt);
+	struct sockaddr_in discGrp;
+	size_t disc_sz;
+	size_t buf_sz = 65535;
+	uint8_t buf[65535] = { 0 };
+
+	if (PIM_DEBUG_AUTORP)
+		zlog_debug("%s: AutoRP sending discovery info", __func__);
+
+	/* Mark true, even if nothing is sent */
+	autorp->mapping_agent_active = true;
+	disc_sz = autorp_build_disc_packet(autorp, buf, buf_sz);
+
+	if (disc_sz > 0) {
+		discGrp.sin_family = AF_INET;
+		discGrp.sin_port = htons(PIM_AUTORP_PORT);
+		inet_pton(PIM_AF, PIM_AUTORP_DISCOVERY_GRP, &discGrp.sin_addr);
+
+		if (setsockopt(autorp->sock, IPPROTO_IP, IP_MULTICAST_TTL,
+			       &(autorp->discovery_scope), sizeof(autorp->discovery_scope)) == 0) {
+			if (setsockopt(autorp->sock, IPPROTO_IP, IP_MULTICAST_IF,
+				       &(autorp->mapping_agent_addrsel.run_addr),
+				       sizeof(autorp->mapping_agent_addrsel.run_addr)) == 0) {
+				if (sendto(autorp->sock, buf, disc_sz, 0,
+					   (struct sockaddr *)&discGrp, sizeof(discGrp)) > 0) {
+					if (PIM_DEBUG_AUTORP)
+						zlog_debug("%s: AutoRP discovery message sent",
+							   __func__);
+				} else if (PIM_DEBUG_AUTORP)
```

Comment:
It will be sent on the configured interval, which defaults to 60 seconds. If sending fails for any reason that will be logged, so I think it's better to leave this guarded so it doesn't spam the logs.